### DISCUSSION TITLE
How to control execution order of MOOSE objects

### DISCUSSION THREAD

**Discussion by NorthMagic** [idaholab/moose#25827] [Q&A General] [answered]
Hello, I'm new to MOOSE, and have a problem on developping a MOOSE application. I applied the Monte Carlo code OpenMC to generate few-group cross sections which used for reactor physics simulations by the MOOSE-based application. The OpenMC was compiled into a dynamic libarary and implemented in MOOSE's material system to generate the few-group cross sections. In an eigenvalue problem, I found that the construction function of the material system derived object were executed by three times, i. e., OpenMC were executed by three times which should have been executed once enough to generated the few-group cross sections. Then, I changed the location of OpenMC routines in the derived object but cannot fix this problem well.

Moreover, I want to compute discontinuous factors of fuel assemblies which needs to sovle the eigenvalue problem and get solutions. The residuals for convergence should be computed from errors between the solutions from MOOSE solver and the reference results from OpenMC. I don't konw how to achieve this iteration loop by using the core framework of MOOSE.

For both of the problems, the key is that I don't know how to control the execution orders of MOOSE objects. Could you give me some instructions?

**Comment by GiudGiud** [comment]
Hello

Materials are executed three times for the local element, the neighbor element and the face value (or something related, I d have to check that one)

I'm not sure we have a way to control that. 

I would think openmc would be run in a user object instead, once for the whole simulation and not once per element?
Is once per element really what you want?

Guillaume

**Reply by NorthMagic** [reply]
Thanks for your answer. As you say, I tried that runing openmc in a user object and getting data from the user object into a material object. In the user object, OpenMC was initialied in _initialize()_ and excuted in _execute()_ derived from the GeneralUserObject. The generated cross sections were got by _computeQpProperties()_ in the material object. But I found that  _computeQpProperties()_ was executed before the user object and changing the execte_on flag in an input file didn't work. What's the problem?

**Reply by GiudGiud** [reply]
You can use Debug/show_execution_order=ALWAYS to see when the user object is being executed

**Reply by YaqiWang** [reply, marked as answer]
That is expected. If you cannot put the openmc call inside the constructor or `initialSetup` of the user object, you will need to manage the `execute_on` parameter very carefully.

**Reply by NorthMagic** [reply]
Thank you very much for your instructions, I fixed this problem by moving openmc routine into a user object's construction function and omitting the override of initialize/execute/finalize.

**Comment by YaqiWang** [comment]
This reference could be helpful to you.
@inproceedings{LaboureDF2018,
  author = {Laboure, Vincent and Ortensi, Javier and Wang, Yaqi and Schunert, Sebastian and DeHart, Mark},
  address = {Cancun, Mexico},
  booktitle = {PHYSOR 2018: Reactor Physics paving the way towards more efficient systems},
  month = {April},
  title = {Iterated Discontinuity Factors Based on Reference Outgoing Currents for Discontinuous Finite Element Diffusion},
  year = {2018}
}

A material is constructed three times for element, face and neighbor face (Advanced topics in https://mooseframework.inl.gov/moose/syntax/Materials/). You probably should use a general user object for calling OpenMC routines as suggested by Guillaume and let the material get the data from the user object.

**Reply by NorthMagic** [reply]
Thanks for your reply, this paper is truly helpful to my work. But what makes me confused mostly is how to implement algorithms of computing DFs by using moose. To solve different types of problem in moose, the _Executioner/type_ is set as Eigenvalue/Steady and so on. In algorithm-1 of the reference paper, an eigenvalue calculation and steady-state iterations need to be done. I don't know how to achieve a complex algorithm by combining and ordering several solving routines based on the moose coding framework. Could you give me some suggestions?

**Reply by YaqiWang** [reply]
DF calculation is not an eigenvalue calculation. With moose, we cast it into a nonlinear problem that has DFs depend on fluxes and DFs affect the solution, i.e. fluxes. The residual `R(phi)` can be represented as `Tr[DF(phi)](phi)`, where Tr is implemented with your transport kernel, same as eigenvalue calculations, except the DGKernel needs to couple in DF. For this "nonlinear" problem, the solution is still fluxes without source, but it can have a non-trivial solution (nonzero solution). Once we obtained the fluxes, we can get DFs. We can use `Steady` executioner to solve this which implements a PJFNK (preconditioned Jcobian-free Newton-Krylov) solver. The initial guess could be tricky but typically a flat initial solution is good for the PJFNK solver converging to the solution.

**Reply by NorthMagic** [reply]
Thank you very much, I will try it.

**Reply by NorthMagic** [reply]
I'm sorry to bother you again, but I'm getting stucked in the implementation of DF calculation for two weeks. The problem is that non-physical (negative and asymmetry) solutions always occur. By checking codes carefully, I found that the error was caused by solving the neutron diffusion equation with `Steady` executioner in which a keff is fixed and comes from a reference solution. But the correct solutions can be obtained by `Eigenvalue` executioner. The difference between the mentioned two calculation is mainly the implementation of `Kernel` for the fission source term in the neutron diffusion equation. I checked it carefully and no mistakes were found in the fission kernel. Could you give me some advices?

**Reply by YaqiWang** [reply]
Your fission kernel must be able to accept a scaling factor for 1/k typically in a postprocessor. With `Eigenvalue`, MOOSE will scale the fission kernel automatically. But with `Steady`, you will need to do the scaling manually. I hope this is the issue.

**Reply by NorthMagic** [reply]
Thank you, I fixed it.
